pkg/recovery/apiserver.go

```diff
@@ -0,0 +1,374 @@
+	if err != nil {
+		return fmt.Errorf("failed to write recovery pod manifest %q: %v", recoveryPodManifestPath, err)
+	}
+
+	// Create client cert
+	ca := crypto.CA{
+		Config:          rootCaConfig,
+		SerialGenerator: &crypto.RandomSerialGenerator{},
+	}
+
+	// Create client certificates for system:admin
+	// (Reuse the serving CA as client CA, this is fine for shortlived localhost recovery apiserver.)
+	clientCert, err := ca.MakeClientCertificateForDuration(&user.DefaultInfo{Name: "system:admin"}, certValidity)
+	if err != nil {
+		return fmt.Errorf("failed to create client certificate: %v", err)
+	}
+
+	clientCertBytes, clientKeyBytes, err := clientCert.GetPEMBytes()
+
+	s.restConfig = &rest.Config{
+		Host: "https://localhost:7443",
+		TLSClientConfig: rest.TLSClientConfig{
+			CAFile:   servingCaCertPath,
+			CertData: clientCertBytes,
+			KeyData:  clientKeyBytes,
+		},
+	}
+
+	kubeconfig, err := s.KubeConfig()
+	if err != nil {
+		return fmt.Errorf("failed to create kubeconfig: %v", err)
+	}
+
+	kubeconfigBytes, err := yaml.Marshal(kubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to marshal kubeconfig: %v", err)
+	}
+
+	kubeconfigPath := filepath.Join(s.recoveryResourcesDir, AdminKubeconfigFileName)
+	err = ioutil.WriteFile(kubeconfigPath, kubeconfigBytes, 600)
+	if err != nil {
+		return fmt.Errorf("failed to write kubeconfig %q: %v", kubeconfigPath, err)
+	}
+
+	return nil
+}
+
+func (s *Apiserver) WaitForHealthz(ctx context.Context) error {
+	// We don't have to worry about connecting to an old, terminating apiserver
+	// as our client certs are unique to this instance
+
+	kubeClientset, err := s.GetKubeClientset()
+	if err != nil {
+		return err
+	}
+
+	return wait.PollUntil(500*time.Millisecond, func() (bool, error) {
+		req := kubeClientset.RESTClient().Get().AbsPath("/healthz")
+
+		klog.V(1).Infof("Waiting for recovery apiserver to come up at %q", req.URL())
+
+		_, err = req.DoRaw()
+		if err != nil {
+			klog.V(5).Infof("apiserver returned error: %v", err)
+			return false, nil
+		}
+
+		return true, nil
+	}, ctx.Done())
+}
+
+func (s *Apiserver) Destroy() error {
+	recoveryPodManifestPath := filepath.Join(s.PodManifestDir, RecoveryPodFileName)
+
+	recoveryPod, err := ReadManifestToV1Pod(recoveryPodManifestPath)
+	if err != nil {
+		return fmt.Errorf("failed to decode file %q: %v", recoveryPodManifestPath, err)
+	}
+
+	resourceDirPath, err := GetVolumeHostPathPath("resource-dir", recoveryPod.Spec.Volumes)
+	if err != nil {
+		return fmt.Errorf("failed to find resource-dir volume for pod manifest %q: %v", recoveryPodManifestPath, err)
+	}
+
+	klog.Infof("Deleting resource-dir %q", resourceDirPath)
+	err = os.RemoveAll(resourceDirPath)
+	if err != nil {
+		return fmt.Errorf("failed to remove recovery pod manifest %q: %v", recoveryPodManifestPath, err)
+	}
+
+	klog.Infof("Deleting recovery pod manifest %q", recoveryPodManifestPath)
+	err = os.Remove(recoveryPodManifestPath)
```

Comment:
this delete should logically happen first

Reply:
> this delete should logically happen first

nm. I understand the indempotency.

Reply:
depends on the logic if it fails in between you leak the dir and second run won't help. Non-working recovery apiserver that you are trying to destroy seems better